### PR TITLE
Added Binary Search Tree [python]

### LINKED ISSUE #344: Binary Search Tree [Python]
#### This issue is about [Algo/DS Name](link to resource for the Algo/DS).

- [x] Issue Title is in form `Algo/DS Name [Language]`.
- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This Algo/DS is worth adding to this repository.
- [x] Assign this issue to me (I want to work on this).
	- [x] I ensure that I am not already assigned to 2 or more issues.
- [x] I will send my PR only after I'm assigned to this issue by any maintainer.
- [x] I understand that I will be unassigned from this issue if I fail to send a PR for this within a week.

### PR DESCRIPTION
**Pull request by Hsankesara**
Fixes #344 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

**Comment by Hsankesara** [comment]
Hey @Anshuman-Verma, I realized that my code showing  ```syntax error``` when compiled with ```python2``` , this is because of ```print_preorder()``` method, but I cannot find out solution of this problem. I want to print every element of tree in a single line using recursion, but unable to do it in both ```python2``` and  ```python3```  simultaneously . @aviaryan @singhpratyush  @aashutoshrathi @Anshuman-Verma, someone please help me here.

File: binary_search_tree/binary_search_tree.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            return False
+
+    def print_preorder(self, node):
+        """
+        Print all element of binary tree in preorder using recurtion
+        """
+        if node is None:
+            return
+        elif node.left is None and node.right is None:
+            print(node.data, end=' ')
```
Comment (aashutoshrathi) [inline binary_search_tree/binary_search_tree.py]: What are you doing with `end=' '` exactly ? @Hsankesara

Comment (Hsankesara) [inline binary_search_tree/binary_search_tree.py]: Nothing just try to print every element of tree in single line...but I changed it now.

File: binary_search_tree/binary_search_tree.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        self.right = None
+
+
+class bst(object):
+
+    def __init__(self):
+        # here root is root of our BST tree which is initially None
+        self.root = None
+
+    def insert_node(self, data):
```
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: Using recursive technique to insert node can reduce the length of the code of this function.

File: binary_search_tree/binary_search_tree.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                                    parent_to_replace = to_replace
+                                    to_replace = to_replace.left
+                                parent_node.right.data = to_replace.data
+                                parent_to_replace.left = to_replace.right
+                        return
+                    else:
+                        iterator = iterator.right
+        return
+
+    def search_node(self, data):
```
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: Here also using a recursive technique can reduce the code length also there won't be any need to check for root separately.

File: binary_search_tree/binary_search_tree.py (hunk 2 above)
Comment (Hsankesara) [inline binary_search_tree/binary_search_tree.py]: Iterative method does not cause any harm. does it?

File: binary_search_tree/binary_search_tree.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            sys.stdout.write(str(node.data) + ' ')
+            self.print_preorder(node.right)
+            if self.root == node:
+                sys.stdout.write('\n')
+            return
+
+
+def main():
+    inp = bst()
+    inp.insert_node(4)
```
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: No need to insert these many nodes, 4-5 should be sufficient.

File: binary_search_tree/binary_search_tree.py (hunk 3 above)
Comment (Hsankesara) [inline binary_search_tree/binary_search_tree.py]: but isn't iterative method faster than recursive?

File: binary_search_tree/binary_search_tree.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                                    parent_to_replace = to_replace
+                                    to_replace = to_replace.left
+                                parent_node.left.data = to_replace.data
+                                parent_to_replace.left = to_replace.right
+                        return
+                    else:
+                        iterator = iterator.left
+                else:
+                    if iterator.right is None:
+                        raise KeyError('Node does not found')
```
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: `Node does not exists`.

File: binary_search_tree/binary_search_tree.py (hunk 2 above)
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: Python codes are supposed to short. That is what makes python different from other languages.

File: binary_search_tree/binary_search_tree.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+                        to_replace = to_replace.left
+                    node.data = to_replace.data
+                    parent_to_replace.left = to_replace.right
+                    self.root = node
+        else:
+            iterator = self.root
+            while iterator is not None:
+                if iterator.data >= data:
+                    if iterator.left is None:
+                        raise KeyError('Node does not found')
```
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: `Node does not exists`.
Edit: It should be `LookupError`

**Comment by Monal5031** [comment]
@Hsankesara Can this implementation be improved by using `search_node()` to return a `node` value along with the `bool` value. The node thus returned be used by `delete_node()` to delete it. This may simplify a lot of work in `delete_node()` which can be done by `search_node()` itself.

**Comment by Hsankesara** [comment]
@Monal5031 you can implement it but then ```search_node``` have to return parent of node you want to delete.But yeah its possible... 😄

**Comment by Monal5031** [comment]
@Hsankesara Ok then please make the changes 😄 accordingly.

File: binary_search_tree/binary_search_tree.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    inp.insert_node(5)
+    inp.insert_node(44)
+    inp.insert_node(69)
+    inp.print_preorder(inp.root)
+    inp.delete_node(69)
+    inp.delete_node(5)
+    inp.delete_node(4)  # delete root
+    inp.print_preorder(inp.root)
+    print(inp.root.data)  # check value of root
+    inp.delete_node(2)  # just to throw exception
```
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: Put this in a `try-except` block.

Comment (Hsankesara) [inline binary_search_tree/binary_search_tree.py]: Whats the problem with ```raise```?

**Comment by Hsankesara** [comment]
yeah @Monal5031 , I didn't use recursive method because I genuinely don't believe using recursion over iteration can improve code quality.

File: binary_search_tree/binary_search_tree.py (hunk 7 above)
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: I knew you'd say that 😆 , Please take a look at your travis build that's the problem.

Comment (Hsankesara) [inline binary_search_tree/binary_search_tree.py]: I saw it, It is due to the exception.

**Comment by Monal5031** [comment]
@Hsankesara If you don't want to go with the pythonic way of doing things I'm okay with that, but I asked one another change that was to use search method for searching the node and using it in delete.

**Comment by Hsankesara** [comment]
I did that change @Monal5031

**Comment by Monal5031** [comment]
@Hsankesara my bad, didn't notice that.

File: binary_search_tree/binary_search_tree.py (hunk 7 above)
Comment (Monal5031) [inline binary_search_tree/binary_search_tree.py]: > Whats the problem with `raise`?

nothing with `raise`, I asked to put this line in `try-except` block due to which the exception will be caught and problem can have exit code 0. I hope I'm clear and correct for you to change this.

File: binary_search_tree/binary_search_tree.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+                        iterator.right = new_node
+                        break
+                    else:
+                        iterator = iterator.right
+        return
+
+    def delete_node(self, data):
+        """
+        Delete element from binary tree
+        Time Complexity: O(Logn)average case and O(n) in worst case
```
Comment (Anshuman-Verma) [inline binary_search_tree/binary_search_tree.py]: Add space around `average`

**Comment by Monal5031** [comment]
@Hsankesara Make necessary changes (i.e. convert iterative methods to recursive) in order to shorten the length of code.

**Comment by Hsankesara** [comment]
@Monal5031 I think you are assigned on this...go on do whatever you want.... 👍

**Comment by Monal5031** [comment]
@Hsankesara I think you didn't read the chats on gitter, I'm assigned to review this PR not make changes.

**Comment by Hsankesara** [comment]
I didn't have electricity for 3 days.So i didn't read gitter chat. I'll make changes when I'll reach gandhinagar which take few days.

**Comment by Anshuman-Verma** [comment]
No problem @Hsankesara

**Comment by Monal5031** [comment]
No worries @Hsankesara just do the changes in time 😄

**Comment by Hsankesara** [comment]
@Monal5031 Why are you not happy with iterative method?

**Comment by Anshuman-Verma** [comment]
I'm not sure whether what approach we should follow, read this [article](http://blog.nextgenetics.net/?e=64) a while ago and got that there are some cons with both approaches. As of with:
**Recursion :**
* Python has a max recursion depth. 
* Excess memory.  

**Iteration :** 
* Need additional checks and conditions (lengthy code).  

But i think we've been following recursion in all implementations so it's better to follow recursion maybe ?
Let @Monal5031 have his views.

**Comment by Monal5031** [comment]
@Hsankesara I think @Anshuman-Verma have wrapped up all the points, but the main issue that remains is that python codes are supposed to short, if something can be done in comparatively less lines with a small performance cost, it should be done. You may see where the python codes can be made more efficient but it increases the lines of code which is not preferred in case of python.

**Comment by Monal5031** [comment]
@Hsankesara Are you going to make the changes or not?

**Comment by Hsankesara** [comment]
I believe I wouldn't @Monal5031

**Pull request closed without merging**